### PR TITLE
Add option to forbid empty test suites

### LINKED ISSUE #4062: Add option to fail if no tests are run
**Is your feature request related to a problem or a nice-to-have?? Please describe.**

At the moment, if `mocha.run()` is called without any tests having been defined, the result is considered a success as the number of failures is 0. In messageformat/messageformat#261 this is causing a problem when Selenium browser tests are being [run in a CI environment](https://travis-ci.org/messageformat/messageformat/jobs/553023787) using [mocha-selenium-bridge](https://github.com/eemeli/mocha-selenium-bridge), and the tests "pass" (note the IE 11 results):

```
▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
Testing Chrome 49.0...
  83 passing (179ms)

    ✓ Chrome 49.0 (8602ms)

▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
Testing Firefox 66.0...
  83 passing (358ms)

    ✓ Firefox 66.0 (10933ms)

▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬
Testing IE 11.0...
  0 passing (0ms)

    ✓ IE 11.0 (6062ms)
```

In our case, this went unnoticed for a number of months as the result was marked as a success. The bug itself was caused by some unsupported-by-IE11 syntax in a test file.

**Describe the solution you'd like**
I would like to have an option to consider "0 passing" tests to be considered an error.

**Describe alternatives you've considered**
I could probably write a custom reporter that would catch this, but that seems like a rather clumsy solution. I'm pretty sure the proper place to fix this would be somewhere [around here](https://github.com/mochajs/mocha/blob/3633a9062458d30d425d743bd0df65c1815b2908/lib/runner.js#L903), but the details of course depend on the implementation.

**Additional context**
I'm willing to submit a PR for this, but I'd rather get feedback first on how e.g. the error code should be handled, and/or whether a separate event would be appropriate for reporters to customise.

### PR DESCRIPTION
**Pull request by eemeli**
### Description of the Change

Adds an option `forbidEmptySuite`, available in the CLI as `--forbid-empty-suite`, that will fail a suite with an error `Empty suite forbidden` if it contains no tests.

### Alternate Designs

A similar result could be achieved by an individual customised reporter, but that seems like the wrong layer at which to address this.

### Why should this be in core?

Because the other options are also defined in the core.

### Benefits

At the moment, if the test definitions fail in a context that does not fail a subsequent `mocha.run()` (e.g. a browser processing multiple `<script>` tags), this failure may not be caught by Mocha, and instead it will erroneously display a positive result with "0 passing" tests. This is problematic in particular in CI tests, where the results are often left unchecked if they "pass".

### Possible Drawbacks

This adds yet another option.

### Applicable issues

Fixes #4062

### CODE REVIEW

**Comment by juergba** [comment]
@eemeli I apalogize for not responding to your PR. Are you still interested?
I would like to review and push it through now.

**Comment by eemeli** [comment]
Yes, I'm still interested in this. Happy to have this reviewed and hopefully merged!

**Comment by juergba** [comment]
What about `pending` tests?
- static: `it.skip()`
- conditionally at runtime: `before(function() {   this.skip();  });` 

We could end up with zero tests run, but the script would still pass.
If we want to cover above scenario, we would have to check the `stats-collector` somehow, at `runner`'s end for the total of tests run.

Some users also create tests dynamically at runtime, when the parsing/creating of the test runner has already been completed.

**Comment by eemeli** [comment]
I would argue that a suite with a skipped test is not "empty", and that the right thing in that case is not to complain. Sure, skipping all tests would therefore result in a suite passing even if none of the tests run, but that'd be intentional in this case.

This logic also allows for using `forbidEmptySuite: true` together with `it.only`, which would otherwise fail if there are multiple suites.

**Comment by eemeli** [comment]
Regarding dynamically generated tests, that's an entirely valid point. I don't think the proposed behaviour would be very surprising in that case, but that should probably still be accounted for. Are there other places than `before()` that might dynamically add tests in an otherwise empty suite? And would the right thing then be to fail before or after the suite-end event is emitted?

**Comment by juergba** [comment]
I don't understand this one:
> This logic also allows for using forbidEmptySuite: true together with it.only [...]

With dynamically generated tests I was refering to scenarios like #4126 etc. You can do this in `it()` and any hook type, but hooks need at least one test in order to run. So the suite wouldn't be empty anyway. 

> And would the right thing then be to fail before or after the suite-end event is emitted?

At suite-end event, I guess.

In your issue you wrote:

> I would like to have an option to consider "0 passing" tests to be considered an error.

In this PR you switched to checking each and every suite, instead of a final test at `runner`'s end. What is your reason?

I'm not sure yet, wether we should go your way, or check once at `runner`'s end.

**Comment by eemeli** [comment]
Perhaps I should rephrase the option to consider it an error if a suite "defines" no tests, rather than "containing" no tests? That might make it altogether clearer.

Regarding `it.only`, I mean the case of having two parallel suites (say, A and B) each of which defines some tests. If one test in B is marked as `it.only`, all of the tests in suite A are skipped. With the current formulation, running those test suites with `--forbid-empty-suite` will pass, but if it's changed to require each suite to contain at least one test that's run, it'll fail.

> With dynamically generated tests I was refering to scenarios like #4126 etc. You can do this in `it()` and any hook type, but hooks need at least one test in order to run. So the suite wouldn't be empty anyway.

Expect for the `before()` hook, right? As far as I can tell, that's normally run even for an empty suite. So it provides a way for an initially empty suite to end up with some tests. And there's probably all sorts of other places, if you're adding tests to other suites than the current one.

> In this PR you switched to checking each and every suite, instead of a final test at `runner`'s end. What is your reason?

Three reasons:
1. Prefer to fail earlier rather than later
2. If suite A contains only suite B, which in turn contains only suite C, make sure that we only throw one error.
3. I presumed that the emptiness of a suite could be determined before it's even run

I'll amend and rebase this PR on top of the latest master, to account for dynamically added tests.

**Comment by eemeli** [comment]
On an entirely different note, I rather appreciate the timing of pointing at an issue filed on the last Friday the 13th, on a Friday the 13th.

**Comment by juergba** [comment]
Yes, and now the corona virus. Will there be any survivors left for using your new feature?

> Expect for the before() hook, right? As far as I can tell, that's normally run even for an empty suite.

No, a `before()` shouldn't run without at least one test.

**Comment by eemeli** [comment]
Well, haven't we all heard recently that testing in particular is rather important just now?

> No, a `before()` shouldn't run without at least one test.

You're right, that was a misunderstanding on my part.

I added a test for a dynamically-defined test in an otherwise empty suite, and it turns out that the current approach already handles [that case](https://github.com/mochajs/mocha/pull/4123/files#diff-c66a17cc5e6589170f15c55ee5b24117) just like it should.

And then I realised that I can use the already-calculated `total` from `suite.grepTotal` to determine if a suite has no tests, and dropped the `suite.isEmpty` method. This means that the complete change required to `runner.runSuite` is [this if statement](https://github.com/mochajs/mocha/pull/4123/files#diff-b3b53682a18f203ac8d29b0e277cad26R735-R737).

And therefore I didn't move the test to the suite-end, because the existing behaviour after those lines prevents the suite-begin and suite-end events from being emitted if the suite is empty, and there's no reason to change the existing behaviour.

I also clarified the docs a little bit, and rebased the branch on the current `master`.

**Comment by juergba** [comment]
@eemeli I will review this weekend.

File: lib/runner.js [hunk 1]
```diff
@@ -732,6 +732,9 @@ Runner.prototype.runSuite = function(suite, fn) {
 
   debug('run suite %s', suite.fullTitle());
 
+  if (!total && this.forbidEmptySuite) {
+    this.fail(suite, new Error('Empty suite forbidden'));
```
Comment (juergba) [inline lib/runner.js]: `Suite` does not extend `Runnable` (as does `Test` and `Hook`), so you can't `fail` a `Suite`.
The reporter output is weird, if the test file contains only an empty suite.
Maybe a dummy test?

**Comment by boneskull** [comment]
@eemeli Are you still interested in working on this?

**Comment by boneskull** [comment]
FWIW I think this is a reasonable addition.  It's not something that can be discovered via static analysis, and people _do_ want to be able to run Mocha w/o this sort of restriction.  

At this point though I wonder if we should add something like a "CI mode" which enables various "strict" behaviors...

**Comment by eemeli** [comment]
Sorry, this fell off my radar. Yes, I'm still interested in working on this. I'll need to review @juergba's comments and get back to you on this.

**Comment by eemeli** [comment]
Closing, as I clearly haven't had the time to work on this.
